rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,270 @@
+
+## Git and NixOS release problems
+
+- The migration will cause merge conflicts for a lot of PRs, but they are trivially resolvable using `git rebase && git push -f` due to Git's file rename tracking.
+- Commits that change moved files in `pkgs/unit` can be cherry-picked to the previous file location without problems for the same reason.
+- `git blame` locally and on GitHub is unaffected, since it follows file renames properly.
+
+## `callPackage` with `nix-build -E`
+
+A commonly recommended way of building package directories in Nixpkgs is to use `nix-build -E 'with import <nixpkgs> {}; callPackage pkgs/applications/misc/hello {}'`.
+Since the path changes `package.nix` is now used, this becomes like `nix-build -E 'with import <nixpkgs> {}; callPackage pkgs/unit/he/hello/package.nix {}'`, which is harder for users.
+However, calling a path like this is an anti-pattern anyway, because it doesn't use the correct Nixpkgs version and it doesn't use the correct argument overrides.
+The correct way of doing it was to add the package to `pkgs/top-level/all-packages.nix`, then calling `nix-build -A hello`.
+This `nix-build -E` workaround is partially motivated by the difficulty of knowing the mapping from attributes to package paths, which is what this RFC improves upon.
+By teaching users that `pkgs/unit/*/<name>` corresponds to `nix-build -A <name>`, the need for such `nix-build -E` workarounds should disappear.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- The existing categorization of packages gets lost. Counter-arguments:
+  - It was never that useful to begin with
+    - The categorization was always incomplete, because packages defined in the language package sets often don't get their own categorized file path.
+    - It was an inconvenient user interface, requiring a checkout or browsing through GitHub
+    - Many packages fit multiple categories, leading to multiple locations to search through instead of one
+  - There's other better ways of discovering similar packages, e.g. [Repology](https://repology.org/)
+- This breaks `builtins.unsafeGetAttrPos "hello" pkgs`. Counter-arguments:
+  - We have to draw a line as to what constitutes the public interface of Nixpkgs. We have decided that making attribute position information part of that is not productive. For context, this information is already accepted to be unreliable at the language level, noting the `unsafe` part of the name.
+  - Support for this could be added to Nix (make `builtins.readDir` propagate file as a position)
+
+# Alternatives
+[alternatives]: #alternatives
+
+TODO: This needs updating
+
+## A different unit base directory
+
+Context: `pkgs/unit` is the unit base directory
+
+Alternatives:
+- Use `unit` (at the Nixpkgs root) instead of `pkgs/unit`.
+  - This is future proof in case we want to make the directory structure more general purpose, but this is out of scope
+- Other name proposals were deemed worse: `pkg`, `component`, `part`, `mod`, `comp`
```

Comment:
**Bikeshedding thread**

"auto" has already been proposed elsewhere, and I'd like to discuss the following alternatives too: "app", "simple".

I like "app" because most of the eligible packages are applications. Of course it might be confusing with libraries, not sure how many packages this affects.

"simple" is meant as a contrast to ecosystem packages like python packages, which make use of additional packaging infrastructure. It is also a hint at most applications in there being "simple" leaf packages

Reply:
Not to say that we were using this definition, or that you referred to these, but in philosophy, the terms "simple" and "unit" have distinct meanings.

"Simple" refers to something that is not composed of parts. In metaphysics, a simple entity is one that has no internal structure or complexity, and is therefore indivisible. Examples of simple entities include mathematical points, atoms, and souls (in some philosophical traditions).

"Unit," on the other hand, refers to a discrete, indivisible entity that is distinct from other entities of the same type. Units are often thought of as basic building blocks or fundamental constituents of reality. For example, in the philosophy of mathematics, numbers are often thought of as units, and more complex mathematical structures can be built up from these units.

So while both "simple" and "unit" refer to indivisible entities, "simple" refers to something that has no internal structure, while "unit" refers to a discrete entity that can be composed into more complex structures.

In this sense, "unit" is more appropriate for our directories.

Reply:
I'm not a big fan of this usage of "leaf" by the way. Without context, it _may_ instead refer to the bootstrap, or the derivation sources. I'd prefer "non-dependency" instead: non-dependency package, non-dependency derivation.

Reply:
Did we already discuss `pkgs/default` as a possible name? Nix already uses "default" for file names as a convention, so why not here too? It hopefully evokes the feeling that this is the "default" folder in `pkgs` and therefore packages should put in there by default unless there are specific reasons not to.